README.md
<p align="center">
  <!--<img alt="GoReleaser Logo" src="https://avatars2.githubusercontent.com/u/24697112?v=3&s=200" height="140" />-->
  <h3 align="center">Delving RAPID</h3>
  <p align="center">Linked Open Data microservice</p>
  <p align="center">
    <a href="https://github.com/delving/rapid/releases/latest"><img alt="Release" src="https://img.shields.io/github/release/delving/rapid.svg?style=flat-square"></a>
    <a href="/LICENSE.md"><img alt="Software License" src="https://img.shields.io/badge/License-Apache%202.0-blue.svg?style=flat-square"></a>
    <a href="https://travis-ci.org/delving/rapid"><img alt="Travis" src="https://img.shields.io/travis/delving/rapid/master.svg?style=flat-square"></a>
    <a href='https://coveralls.io/github/delving/rapid?branch=develop'><img src='https://coveralls.io/repos/github/delving/rapid/badge.svg?branch=develop' alt='Coverage Status' /></a>

    <a href="https://goreportcard.com/report/github.com/delving/rapid"><img alt="Go Report Card" src="https://goreportcard.com/badge/github.com/delving/rapid?style=flat-square"></a>
    <!--<a href="http://godoc.org/github.com/delving/rapid"><img alt="Go Doc" src="https://img.shields.io/badge/godoc-reference-blue.svg?style=flat-square"></a>-->
    <a href="https://github.com/goreleaser"><img alt="Powered By: GoReleaser" src="https://img.shields.io/badge/powered%20by-goreleaser-green.svg?style=flat-square"></a>
  </p>
</p>


RAPID is a recursive acronym that stands for *Rapid API Delving*. 

The goal of RAPID is to provide *an API Framework that makes it easy and predictable for webdevelopers to work with arbitrarily structured RDF and leverage semantic network technology*.

The core functionality that it aims to provide can be summarised by the acronym *SILAS*:

* **S**PARQL proxy
* **I**ndex RDF
* **L**inked Open Data Resolver
* **A**ggregate and transform RDF
* **S**earch RDF

Part of the design is to require as little external dependencies outside the compiled *Golang* binary as possible. 

## Project Status

Rapid v1.0 will be released in July 2018. Check the [Changelog] for the full details.

[Changelog]:https://github.com/delving/rapid/blob/master/CHANGELOG.md

## Table of Contents

* [Getting Started](#getting-started)
* [Developing](#developing)
    - [Branching Model](#branching-model)
    - [Dependency Management](#dependency-management)
* [License](#license)

## Getting Started 

RAPID is written in Golang, so you have to setup your Golang environment first, see [Golang Installation].

After that you can glone it from bitbucket:

    $ git clone git@github.com:delving/rapid.git $GOPATH/src/github.com/delving

Or use `go get`

    $ go get github.com/delving/rapid

In order to get the dependencies, install [dep]:

    $ curl https://raw.githubusercontent.com/golang/dep/master/install.sh | sh
    $ dep ensure

Start the server with the default configuration.

    $ rapid http

## Developing

### Branching Model

We use the [GitFlow](https://github.com/nvie/gitflow) branching model. Instead of merging features and releases locally, you should use `git flow feature publish <your feature>` and then create a pull-request.

### Dependency Management

We have decided to use [dep] for vendoring. All pinned dependencies are stored in the `./vendor` directory.

Here are some basic commands to work with dep.

Install the dependencies and revisions listed in the lock file into the vendor directory. If no lock file exists an update is run.

    $ dep ennsure

Install the latest dependencies into the vendor directory matching the version resolution information. The complete dependency tree is installed. A lock file is created from the final output.

    $ dep update

Add a new dependency to the `Gopkg.toml`, install the dependency, and re-resolve the dependency tree. Optionally, put a version after an anchor.

    $ dep ensure --add github.com/foo/bar

or 

    $ dep ensure --add github.com/foo/bar#^1.2.3

### System Dependencies

Currently, RAPID depends on a few external dependencies.

- A search engine: We currently support the [Apache Lucene]-based search-engine [ElasticSearch] 5.
- A triple store (optional): We currently support [Apache Fuseki] out of the box. But any Triple store with a [SPARQL 1.1] and [SPARQL Update 1.1] endpoint should work.

Current development and production uses the following versions:

- [ElasticSearch] 5.6 (configured port: 9200)
- [Apache Fuseki] 3.5.0 (configured port: 3030)

All of these dependencies are available on the major platforms via package-managers or direct download: Linux, MacOS and Windows. Please follow the respective installation instructions to get them up and running. RAPID will complain at startup when these are not available.

The easiest setup for development is using the supplied `docker-compose.yml`. Run the following command to start all the external dependencies. 

    `$ docker-compose up`

It also includes [Kibana] also runs on `http://localhost:5601` to provide a development console for ElasticSearch.

### Please help out

This project is still under development. Feedback and suggestions are very
welcome and we encourage you to use the [Issues
list](http://github.com/delving/rapid/issues) on Github to provide that
feedback.

Feel free to fork this repo and to commit your additions. For a list of all
contributors, please see the [AUTHORS](AUTHORS) file.

### Contributing

Fork the repository.  Then, run:

    git clone --recursive git@github.com:<username>/rapid.git
    cd rapid 
    git branch master origin/master
    git flow init -d
    git flow feature start <your feature>

Then, do work and commit your changes.  

    git flow feature publish <your feature>

When done, open a pull request to your feature branch.


## License

Copyright (c) 2017-present Delving B.V.

Licensed under [Apache 2.0](./License)

[Golang Installation]: https://golang.org/doc/install
[dep]: https://golang.github.io/dep/ 
[Kibana]: https://www.elastic.co/products/kibana
[golang]: https://golang.org/
[Apache Lucene]: https://lucene.apache.org/ 
[Apache Fuseki]: https://jena.apache.org/documentation/fuseki2/index.html
[ElasticSearch]: https://www.elastic.co/guide/en/elasticsearch/reference/6.2/getting-started.html
[SPARQL 1.1]: https://www.w3.org/TR/sparql11-query/
[SPARQL Update 1.1]: https://www.w3.org/TR/sparql11-update/


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

[Unreleased]: https://github.com/delving/rapid/HEAD


config/config_test.go
// Copyright © 2017 Delving B.V. <info@delving.eu>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package config_test

import (
	. "github.com/delving/rapid/config"

	. "github.com/onsi/ginkgo"
	. "github.com/onsi/gomega"
)

var _ = Describe("Config", func() {

	Describe("after initialisation", func() {

		Context("when calling initConfig", func() {

			It("should be initialised with defaults", func() {
				InitConfig()
				Expect(Config.HTTP.Port).To(Equal(3001))
				//Expect(Config.OrgID).ToNot(BeEmpty())
			})

		})
	})

	Describe("build information", func() {

		Context("when initializing", func() {

			It("should set devBuild  on empty", func() {
				info := NewBuildVersionInfo(
					"", "", "ginkgo", "",
				)
				Expect(info).ToNot(BeNil())
				Expect(info.Version).To(Equal("devBuild"))
				Expect(info.BuildAgent).To(Equal("ginkgo"))
			})

			It("should pretty print to JSON", func() {
				info := NewBuildVersionInfo(
					"", "", "ginkgo", "",
				)
				Expect(info).ToNot(BeNil())
				json, err := info.JSON(true)
				Expect(err).ToNot(HaveOccurred())
				Expect(json).To(ContainSubstring("ginkgo"))
			})
		})
	})

})


config/namespace_test.go
// Copyright © 2017 Delving B.V. <info@delving.eu>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package config_test

import (
	c "github.com/delving/rapid/config"

	. "github.com/onsi/ginkgo"
	. "github.com/onsi/gomega"
)

var _ = Describe("Namespace", func() {

	Describe("Has a NameSpaceMap", func() {

		Context("When creating a New NameSpaceMap", func() {

			It("Should create a Map", func() {
				nsMap := c.NewNameSpaceMap()
				Expect(nsMap).ToNot(BeNil())
			})
		})

		Context("When adding a key to the NamespaceMap", func() {

			It("should have no items", func() {
				nsMap := c.NewNameSpaceMap()
				prefix2base, base2prefix := nsMap.Len()
				Expect(prefix2base).To(Equal(0))
				Expect(prefix2base).To(Equal(base2prefix))
			})

			It("Should add the key Map", func() {
				nsMap := c.NewNameSpaceMap()
				nsMap.Add("dc", "http://purl.org/dc/elements/1.1/")
				Expect(nsMap).ToNot(BeNil())
				prefix2base, base2prefix := nsMap.Len()
				Expect(prefix2base).To(Equal(1))
				Expect(prefix2base).To(Equal(base2prefix))
			})

			It("Should not add the key twice", func() {
				nsMap := c.NewNameSpaceMap()
				nsMap.Add("dc", "http://purl.org/dc/elements/1.1/")
				nsMap.Add("dc", "http://purl.org/dc/elements/1.1/")
				Expect(nsMap).ToNot(BeNil())
				prefix2base, base2prefix := nsMap.Len()
				Expect(prefix2base).To(Equal(1))
				Expect(prefix2base).To(Equal(base2prefix))
			})
		})

		Context("When retrieving from the NameSpaceMap", func() {

			It("should return not ok when a key is not found", func() {
				nsMap := c.NewNameSpaceMap()
				base, ok := nsMap.GetBaseURI("dc")
				Expect(base).To(BeEmpty())
				Expect(ok).To(BeFalse())
				prefix, ok := nsMap.GetPrefix("http://purl.org/dc/elements/1.1/")
				Expect(prefix).To(BeEmpty())
				Expect(ok).To(BeFalse())
			})

			It("should return ok when the key is found", func() {
				nsMap := c.NewNameSpaceMap()
				nsMap.Add("dc", "http://purl.org/dc/elements/1.1/")
				base, ok := nsMap.GetBaseURI("dc")
				Expect(base).ToNot(BeEmpty())
				Expect(ok).To(BeTrue())
				prefix, ok := nsMap.GetPrefix("http://purl.org/dc/elements/1.1/")
				Expect(prefix).ToNot(BeEmpty())
				Expect(ok).To(BeTrue())
			})
		})

		Context("When deleting a key", func() {

			It("should remove the key from the prefix map", func() {
				nsMap := c.NewNameSpaceMap()
				nsMap.Add("dc", "http://purl.org/dc/elements/1.1/")
				prefix2base, base2prefix := nsMap.Len()
				Expect(prefix2base).To(Equal(1))
				Expect(prefix2base).To(Equal(base2prefix))
				nsMap.DeletePrefix("dc")
				prefix2base, base2prefix = nsMap.Len()
				Expect(prefix2base).To(Equal(0))
				Expect(prefix2base).To(Equal(base2prefix))
			})

			It("should remove the key from the prefix map", func() {
				nsMap := c.NewNameSpaceMap()
				nsMap.Add("dc", "http://purl.org/dc/elements/1.1/")
				prefix2base, base2prefix := nsMap.Len()
				Expect(prefix2base).To(Equal(1))
				Expect(prefix2base).To(Equal(base2prefix))
				nsMap.DeleteBaseURI("http://purl.org/dc/elements/1.1/")
				prefix2base, base2prefix = nsMap.Len()
				Expect(prefix2base).To(Equal(0))
				Expect(prefix2base).To(Equal(base2prefix))
			})
		})

		Context("When initialised from Config", func() {

			It("Should Contain the same number of entries as the config list", func() {

			})
		})

		Context("When setting defaults", func() {

			It("should set all the default namespaces", func() {
				nsMap := c.NewNameSpaceMap()
				Expect(nsMap).ToNot(BeNil())
				k, _ := nsMap.Len()
				Expect(k).To(Equal(0))
				nsMap.SetDefaultNameSpaces()
				k, v := nsMap.Len()
				Expect(k).To(Equal(v))
				Expect(k).To(Equal(26))
			})
		})

	})

	Describe("Should be able to deal with namespace from a uri", func() {

		Context("When given an URI as a string", func() {

			It("Should split when given an URI with a #", func() {
				rdfType := "http://www.w3.org/1999/02/22-rdf-syntax-ns#type"
				base, name := c.SplitURI(rdfType)
				Expect(name).To(Equal("type"))
				Expect(base).To(HaveSuffix("#"))
			})

			It("Should split when given an URI with a #", func() {
				dcSubject := "http://purl.org/dc/elements/1.1/subject"
				base, name := c.SplitURI(dcSubject)
				Expect(name).To(Equal("subject"))
				Expect(base).To(HaveSuffix("/"))
			})
		})

		Context("when given a URI", func() {

			nsMap := c.NewNameSpaceMap()
			nsMap.Add("dc", "http://purl.org/dc/elements/1.1/")

			It("should return the search label", func() {
				dcSubject := "http://purl.org/dc/elements/1.1/subject"
				label, err := nsMap.GetSearchLabel(dcSubject)
				Expect(err).ToNot(HaveOccurred())
				Expect(label).ToNot(BeEmpty())
				Expect(label).To(Equal("dc_subject"))
			})

			It("should throw an error when namespace is not found", func() {
				dcSubject := "http://purl.org/dc/elements/1.3/subject"
				_, err := nsMap.GetSearchLabel(dcSubject)
				Expect(err).To(HaveOccurred())
			})
		})

	})
})


config/config_suite_test.go
// Copyright © 2017 Delving B.V. <info@delving.eu>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package config_test

import (
	. "github.com/onsi/ginkgo"
	. "github.com/onsi/gomega"

	"testing"
)

func TestConfig(t *testing.T) {
	RegisterFailHandler(Fail)
	RunSpecs(t, "Config Suite")
}


main.go
// Copyright © 2017 Delving B.V. <info@delving.eu>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import "github.com/delving/rapid/cmd"

var (
	// Version of the application. (Injected at build time)
	Version = "0.1.0-SNAPSHOT"
	// BuildStamp is the timestamp of the application. (Injected at build time)
	BuildStamp = "1970-01-01 UTC"
	// BuildAgent is the agent that created the current build. (Injected at build time)
	BuildAgent string
	// GitHash of the current build. (Injected at build time.)
	GitHash string
)

func main() {
	cmd.Execute(Version, BuildStamp, BuildAgent, GitHash)
}


cmd/http.go
// Copyright © 2017 Delving B.V. <info@delving>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package cmd

import (
	"fmt"

	"github.com/spf13/cobra"
	"github.com/spf13/viper"
)

// httpCmd represents the http command
var httpCmd = &cobra.Command{
	Use:   "http",
	Short: "Start the WebServer process",
	Long: `Starting the webserver https process.

You can find an overview of available endpoints on https://localhost:443/endpoints`,
	Run: func(cmd *cobra.Command, args []string) {
		fmt.Println("Starting server")
		//server.Start(buildInfo)
	},
}

func init() {
	RootCmd.AddCommand(httpCmd)

	httpCmd.Flags().IntP("port", "p", 3001, "Port to run Application server on")
	viper.BindPFlag("port", httpCmd.Flags().Lookup("port"))
}


config/namespace.go
// Copyright © 2017 Delving B.V. <info@delving.eu>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package config

import (
	"fmt"
	"strings"
	"sync"
)

// NameSpace is a container for Namespaces base URLs and prefixes
// This is used by resolving namespaces in the RDF conversions
type NameSpace struct {
	Base   string `json:"base"`
	Prefix string `json:"prefix"`
}

// NameSpaceMap contains all the namespaces
type NameSpaceMap struct {
	sync.RWMutex
	prefix2base map[string]string
	base2prefix map[string]string
}

// NewNameSpaceMap creates a new NameSpaceMap
func NewNameSpaceMap() *NameSpaceMap {
	return &NameSpaceMap{
		prefix2base: make(map[string]string),
		base2prefix: make(map[string]string),
	}
}

// Len counts the number of keys in the Map
func (n *NameSpaceMap) Len() (int, int) {
	return len(n.prefix2base), len(n.base2prefix)
}

// Add adds a namespace to the namespace Map
func (n *NameSpaceMap) Add(prefix, base string) {
	n.Lock()
	n.prefix2base[prefix] = base
	n.base2prefix[base] = prefix
	n.Unlock()
}

// AddNameSpace is a convenience function to add NameSpace objects to the Map
func (n *NameSpaceMap) AddNameSpace(ns NameSpace) {
	n.Add(ns.Prefix, ns.Base)
}

//// Load loads the namespaces from the config object
//func (n *NameSpaceMap) Load(c *RawConfig) {
//for _, ns := range c.NameSpaces {
//n.AddNameSpace(ns)
//}
//}

//// NewConfigNameSpaceMap creates a map from the NameSpaces defined in the config
//func NewConfigNameSpaceMap(c *RawConfig) *NameSpaceMap {
//nsMap := NewNameSpaceMap()
//nsMap.SetDefaultNameSpaces()
//nsMap.Load(c)
//return nsMap
//}

// GetBaseURI returns the base URI from the prefix
func (n *NameSpaceMap) GetBaseURI(prefix string) (base string, ok bool) {
	n.RLock()
	base, ok = n.prefix2base[prefix]
	n.RUnlock()
	return base, ok
}

// GetPrefix returns the prefix for a base URI
func (n *NameSpaceMap) GetPrefix(baseURI string) (prefix string, ok bool) {
	n.RLock()
	prefix, ok = n.base2prefix[baseURI]
	n.RUnlock()
	return prefix, ok
}

// DeletePrefix removes a namespace from the NameSpaceMap
func (n *NameSpaceMap) DeletePrefix(prefix string) {
	n.Lock()
	base, ok := n.prefix2base[prefix]
	if ok {
		delete(n.base2prefix, base)
	}
	delete(n.prefix2base, prefix)
	n.Unlock()
}

// DeleteBaseURI removes a namespace from the NameSpaceMap
func (n *NameSpaceMap) DeleteBaseURI(base string) {
	n.Lock()
	prefix, ok := n.base2prefix[base]
	if ok {
		delete(n.prefix2base, prefix)
	}
	delete(n.base2prefix, base)
	n.Unlock()
}

// ByPrefix returns the map with prefixes as keys
func (n *NameSpaceMap) ByPrefix() map[string]string {
	return n.prefix2base
}

// SplitURI takes a given URI and splits it into a base URI and a local name
func SplitURI(uri string) (base string, name string) {
	index := strings.LastIndex(uri, "#") + 1

	if index > 0 {
		return uri[:index], uri[index:]
	}

	index = strings.LastIndex(uri, "/") + 1

	if index > 0 {
		return uri[:index], uri[index:]
	}

	return "", uri
}

// GetSearchLabel returns the search label for a Predicate URI
func (n *NameSpaceMap) GetSearchLabel(uri string) (string, error) {
	base, label := SplitURI(uri)
	prefix, ok := n.GetPrefix(base)
	if !ok {
		return "", fmt.Errorf("no prefix found in ns map for %s + %s", base, label)
	}
	return fmt.Sprintf("%s_%s", prefix, label), nil
}

var defaultNameSpaces = map[string]string{
	"abc":         "http://www.ab-c.nl/",
	"abm":         "http://purl.org/abm/sen",
	"cc":          "http://creativecommons.org/ns#",
	"custom":      "http://www.delving.eu/namespaces/custom/",
	"dbpedia-owl": "http://dbpedia.org/ontology/",
	"dc":          "http://purl.org/dc/elements/1.1/",
	"dcterms":     "http://purl.org/dc/terms/",
	"delving":     "http://schemas.delving.eu/",
	"devmode":     "http://localhost:8000/resource/",
	"edm":         "http://www.europeana.eu/schemas/edm/",
	"europeana":   "http://www.europeana.eu/schemas/ese/",
	"ebucore":     "urn:ebu:metadata-schema:ebuCore_2014/",
	"foaf":        "http://xmlns.com/foaf/0.1/",
	"gn":          "http://www.geonames.org/ontology#",
	"icn":         "http://www.icn.nl/schemas/icn/",
	"narthex":     "http://schemas.delving.eu/narthex/terms/",
	"nave":        "http://schemas.delving.eu/nave/terms/",
	"ore":         "http://www.openarchives.org/ore/terms/",
	"owl":         "http://www.w3.org/2002/07/owl#",
	"raw":         "http://delving.eu/namespaces/raw/",
	"rda":         "http://rdvocab.info/ElementsGr2/",
	"rdf":         "http://www.w3.org/1999/02/22-rdf-syntax-ns#",
	"rdfs":        "http://www.w3.org/2000/01/rdf-schema#",
	"skos":        "http://www.w3.org/2004/02/skos/core#",
	"tib":         "http://schemas.delving.eu/resource/ns/tib/",
	"wgs84_pos":   "http://www.w3.org/2003/01/geo/wgs84_pos#",
}

// SetDefaultNameSpaces sets the default namespaces that are supported
func (n *NameSpaceMap) SetDefaultNameSpaces() {
	for prefix, baseURI := range defaultNameSpaces {
		n.Add(prefix, baseURI)
	}
}


config/multitenancy.go
// Copyright © 2017 Delving B.V. <info@delving.eu>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package config

import (
	"fmt"

	"github.com/spf13/viper"
)

// TenantConfig holds the configuration for each enabled organisation
// This data is loaded from support TOML files
type Tenant struct {
	OrgID string `json:"orgID"` // organisation ID
	LOD   `json:"lod"`
}

// LOD holds all the configuration for the Linked Open Data (LOD) functionality
type LOD struct {
	// make the lod endpoint available
	Enabled bool `json:"enabled"`

	// the 303 redirect entry point.  This is where the content negotiation happens
	Resource string `json:"resource"`

	// the endpoint that renders the data as formatted HTML
	HTML string `json:"html"`

	// the endpoint that renders the RDF data in the requested RDF format. Currently, JSON-LD and N-triples are supported
	RDF string `json:"rdf"`

	// the regular expression to convert the subject uri to the uri for the external Page view
	HTMLRedirectRegex string `json:"redirectregex"`
}

// NewTenant creates a new instance of the Tenant
func NewTenant(cfg *viper.Viper) (*Tenant, error) {
	// setDefaults
	setTenantDefaults(cfg)

	// create tenant
	var t Tenant
	err := cfg.Unmarshal(&t)
	if err != nil {
		return nil, err
	}
	if t.OrgID == "" {
		return nil, fmt.Errorf("unable to find Tenant Configuration.\n")
	}
	return &t, nil
}

// setTenantDefaults sets the sub-tenants defaults
func setTenantDefaults(cfg *viper.Viper) {

	// lod
	cfg.SetDefault("LOD.enabled", true)
	cfg.SetDefault("LOD.html", "page")
	cfg.SetDefault("LOD.rdf", "data")
	cfg.SetDefault("LOD.resource", "resource")
	cfg.SetDefault("LOD.redirectregex", "")
}


config/config.go
// Copyright © 2017 Delving B.V. <info@delving.eu>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package config

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"strings"

	homedir "github.com/mitchellh/go-homedir"
	"github.com/spf13/viper"
)

var (
	// Config is the general configuration object
	Config RapidConfig

	// CfgFile is the path to the config file
	CfgFile string
)

func init() {
	// make sure the config is initialised first
	InitConfig()
}

// RapidConfig holds all the configuration blocks.
// These are bound from cli, Environment variables or configuration files by
// Viper.
type RapidConfig struct {
	WildcardTenant string   `json:"wildcardOrg"`
	EnabledTenants []string `json:"enableOrgIDs"`
	HTTP           `json:"http"`
	ElasticSearch  `json:"elasticsearch"`
	DevMode        bool `json:"devmode"`
	Tenants        map[string]*Tenant
}

// ElasticSearch holds all the configuration values
// It is bound by Viper.
type ElasticSearch struct {
	Urls        []string `json:"urls"`
	IndexName   string   `json:"index"`
	Proxy       bool     `json:"proxy"`
	EnableV1    bool     `json:"enableV1"` // Enable v1 style index
	EnableTrace bool     `json:"enableTrace"`
}

// HTTP holds all the configuration for the http server subcommand
type HTTP struct {
	Port int `json:"port" mapstructure:"port"`
}

// InitConfig reads in config file and ENV variables if set.
func InitConfig() {
	if CfgFile != "" {
		// Use config file from the flag.
		viper.SetConfigFile(CfgFile)
	} else {
		// Find home directory.
		home, err := homedir.Dir()
		if err != nil {
			fmt.Println(err)
			os.Exit(1)
		}

		// Search config in home directory with name "rapid" (without extension).
		viper.AddConfigPath("/etc/default/")
		viper.AddConfigPath(home)
		viper.AddConfigPath(".")
		viper.SetConfigName("rapid")
	}

	viper.SetEnvPrefix("RAPID")
	viper.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))
	viper.AutomaticEnv() // read in environment variables that match

	setDefaults()

	// If a config file is found, read it in.
	if err := viper.ReadInConfig(); err == nil {
		log.Printf("Using config file: %s", viper.ConfigFileUsed())
	}
	err := viper.Unmarshal(&Config)
	if err != nil {
		log.Fatal(
			fmt.Sprintf("unable to decode into struct, %v", err),
		)
	}

	//for _, tenantID := range Config.EnabledTenants {
	//tenantConfig := viper.Sub(tenantID)
	//_, err := NewTenant(tenantConfig)
	//if err != nil {
	//log.Fatal(
	//fmt.Sprintf("Unable to decode subTenant\n", tenantID),
	//)
	//}
	//Config.Tenants[tenantID] = tenant
	//}

}

// set the RapidConfig default values
func setDefaults() {

	// setting defaults
	viper.SetDefault("HTTP.port", 3001)
	viper.SetDefault("WildcardTenant", "all")
	viper.SetDefault("EnabledTenants", []string{"all", "default"})
	viper.SetDefault("DevMode", false)

	// elastic
	viper.SetDefault("ElasticSearch.urls", []string{"http://localhost:9200"})
	viper.SetDefault("ElasticSearch.enabled", true)
	viper.SetDefault("ElasticSearch.IndexName", "hub3-rapid")
	viper.SetDefault("ElasticSearch.Proxy", false)
	viper.SetDefault("ElasticSearch.IndexV1", false)
	viper.SetDefault("ElasticSearch.EnableTrace", false)
}

// BuildVersionInfo holds all the version information
type BuildVersionInfo struct {
	Version    string `json:"version"`
	Commit     string `json:"commit"`
	BuildAgent string `json:"buildAgent"`
	BuildDate  string `json:"buildDate"`
}

// NewBuildVersionInfo creates a BuildVersionInfo struct
func NewBuildVersionInfo(version, commit, buildagent, builddate string) *BuildVersionInfo {
	if version == "" {
		version = "devBuild"
	}
	return &BuildVersionInfo{
		Version:    version,
		Commit:     commit,
		BuildAgent: buildagent,
		BuildDate:  builddate,
	}
}

// JSON returns a json version of the BuildVersionInfo
func (b BuildVersionInfo) JSON(pretty bool) ([]byte, error) {
	if pretty {
		return json.MarshalIndent(b, "", "\t")
	}
	return json.Marshal(b)
}
